compiler/grainformat/reformat.re

```diff
@@ -2114,396 +2930,300 @@ and print_expression =
+                switch (optexpression3) {
+                | Some(expr) =>
+                  Doc.concat([
+                    switch (expr.pexp_desc) {
+                    | PExpBlock(_) => Doc.space
+                    | _ => Doc.line
+                    },
+                    Doc.group(
+                      print_expression(
+                        ~original_source,
+                        ~comments=comments_before_loop_expression,
+                        expr,
                       ),
-                    ])
-                  | None => Doc.nil
-                  },
-                ]),
+                    ),
+                  ])
+                | None => Doc.nil
+                },
               ]),
             ),
             Doc.softLine,
             Doc.rparen,
           ]),
         ),
         Doc.space,
-        print_expression(
-          ~parentIsArrow=false,
-          ~endChar=None,
-          ~original_source,
-          ~parent_loc,
-          expression4,
+        Doc.group(
+          print_expression(
+            ~original_source,
+            ~comments=comments_in_expression4,
+            expression4,
+          ),
         ),
-      ])
-    | PExpContinue => Doc.group(Doc.concat([Doc.text("continue")]))
-    | PExpBreak => Doc.group(Doc.concat([Doc.text("break")]))
+      ]);
+    | PExpContinue => Doc.text("continue")
+    | PExpBreak => Doc.text("break")
     | PExpConstraint(expression, parsed_type) =>
+      let comments_in_expression =
+        Comment_utils.get_comments_inside_location(
+          ~location=expression.pexp_loc,
+          comments,
+        );
+
       Doc.group(
         Doc.concat([
           print_expression(
-            ~parentIsArrow=false,
-            ~endChar=None,
             ~original_source,
-            ~parent_loc,
+            ~comments=comments_in_expression,
             expression,
           ),
-          Doc.text(":"),
-          Doc.space,
-          Doc.indent(
-            Doc.concat([
-              Doc.softLine,
-              Doc.lparen, // TODO needed to fix compiler bug (trailing type annotation needs paren, #866)
-              print_type(parsed_type, original_source),
-              Doc.rparen,
-            ]),
-          ),
-        ]),
-      )
-    | PExpLambda(patterns, expression) =>
-      let args =
-        if (List.length(patterns) == 0) {
-          Doc.concat([Doc.lparen, Doc.rparen]);
-        } else if (List.length(patterns) > 1) {
-          Doc.group(
-            Doc.indent(
+          switch (parsed_type.ptyp_desc) {
+          | PTyConstr(locidentifier, parsedtypes) =>
+            switch (parsedtypes) {
+            | [] =>
               Doc.concat([
-                Doc.softLine,
-                Doc.lparen,
+                Doc.text(":"),
+                Doc.space,
+                print_type(
+                  ~original_source,
+                  ~comments,
+                  ~trailing_separator=false,
```

Comment:
I want to look at this argument—is it worth it to have this flag or can we just put that code in the paths where it is needed?

Reply:
Turns out we don't need it at all!